raiden/network/rpc/client.py

```diff
@@ -77,20 +79,57 @@ def patch_send_transaction(client, nonce_offset=0):
         client.call('eth_nonce', encode_hex(client.sender), 'pending')
     except:
         patch_necessary = True
+        client.last_nonce_update = 0
+        client.current_nonce = None
+        client.nonce_lock = Semaphore()
 
     def send_transaction(sender, to, value=0, data='', startgas=GAS_LIMIT,
                          gasprice=GAS_PRICE, nonce=None):
         """Custom implementation for `pyethapp.rpc_client.JSONRPCClient.send_transaction`.
         This is necessary to support other remotes that don't support pyethapp's extended specs.
         @see https://github.com/ethereum/pyethapp/blob/develop/pyethapp/rpc_client.py#L359
         """
-        pending_transactions_hex = client.call(
-            'eth_getTransactionCount',
-            address_encoder(sender),
-            'pending',
-        )
-        pending_transactions = int(pending_transactions_hex, 16)
-        nonce = pending_transactions + nonce_offset
+        def get_nonce():
```

Comment:
The logic of `get_nonce` seems sound, but I had a hard time reading it, these are my suggestions:

- "expand" the code from `_query_nonce` and `get_nonce` inside `send_transaction`
  - alternative: remove `get_nonce` and make the code for initialization it's own stand alone function.
- use the lock as a context manager